Sources/MongoSwift/BSON/Document+Sequence.swift

```diff
@@ -0,0 +1,121 @@
+extension Document: Sequence {
+    /// Returns a `DocumentIterator` over the values in this `Document`. 
+    public func makeIterator() -> DocumentIterator {
+        guard let iter = DocumentIterator(forDocument: self) else {
+            preconditionFailure("Failed to initialize an iterator over document \(self)")
+        }
+        return iter
+    }
+}
+
+/// An iterator over the values in a `Document`. 
+public class DocumentIterator: IteratorProtocol {
+    /// the libbson iterator. it must be a `var` because we use it as
+    /// an inout argument
+    internal var iter: bson_iter_t
+
+    /// Initializes a new iterator over the contents of `doc`. `doc` must remain alive for 
+    /// the lifetime of the iterator. Returns `nil` if the key is not found, or if an iterator 
+    /// cannot be created over `doc` due to an error from e.g. corrupt data.
+    internal init?(forDocument doc: Document) {
+        self.iter = bson_iter_t()
+        if !bson_iter_init(&self.iter, doc.data) { return nil }
+    }
+
+    /// Initializes a new iterator over the contents of `doc`. `doc` must remain alive for the
+    /// lifetime of the iterator. Returns `nil` if an iterator cannot be created over `doc` due
+    /// to an error from e.g. corrupt data, or if the key is not found.
+    internal init?(forDocument doc: Document, advancedTo key: String) {
+        self.iter = bson_iter_t()
+        if !bson_iter_init_find(&iter, doc.data, key.cString(using: .utf8)) {
+            return nil
+        }
+    }
+
+    /// Advances the iterator forward one value. Returns false if there is an error moving forward
+    /// or if at the end of the document. Returns true otherwise.
+    private func advance() -> Bool {
+        return bson_iter_next(&self.iter)
+    }
+
+    /// Returns the current key. Assumes the iterator is in a valid position.
+    internal var currentKey: String {
+        return String(cString: bson_iter_key(&self.iter))
+    }
+
+    /// Returns the current value. Assumes the iterator is in a valid position.
+    internal var currentValue: BsonValue? {
+        do {
+            switch self.currentType {
+            case .symbol:
+                return try Symbol.asString(from: self)
+            case .dbPointer:
+                return try DBPointer.asDocument(from: self)
+            default:
+                return try DocumentIterator.BsonTypeMap[currentType]?.init(from: self)
+            }
+        } catch {
+            preconditionFailure("Error getting current value from iterator: \(error)")
+        }
+    }
+
+    /// Returns the current value's type. Assumes the iterator is in a valid position.
+    internal var currentType: BsonType {
+        return BsonType(rawValue: bson_iter_type(&iter).rawValue) ?? .invalid
+    }
+
+    /// Returns the keys from the iterator's current position to the end.
+    internal var keys: [String] {
+        var keys = [String]()
+        while self.advance() { keys.append(self.currentKey) }
```

Comment:
Is this accessor supposed to permanently advance the iterator, or should you revert to the previous position before returning? Likewise for `values` below.

Reply:
If these are internal and only used on a disposable iterator, I don't think there's a problem. Perhaps a comment mentioning that they leave the iterator exhausted would be helpful, though.

I think resetting the iterator would entail stashing the current key and then later resetting the position back to the original key; although, documents can have duplicate keys so that's probably not fool-proof.